### DISCUSSION TITLE
Error when compiling Mohr-Coulomb Mfront files from the tutorial in MFront gallery

### DISCUSSION THREAD

**Discussion by gitable05** [thelfer/tfel#79] [Q&A] [answered]
Hi, 

This question is first [posted in FEniCS discourse](https://fenicsproject.discourse.group/t/error-when-compiling-mohr-coulomb-mfront-file-examples-from-mfront-gallery/7467) and it was suggested that I post it here. After the day of posting, I sort of found a solution, but I would still like to bring this to your attention. Thanks!

--

I am interested in doing simulations in FEniCS with Mohr-Coulomb criterion. I followed this [tutorial 1](https://comet-fenics.readthedocs.io/en/latest/demo/plasticity_mfront/plasticity_mfront.py.html) which implements a von Mises criterion with MFront. Here, I successfully compiled the MFront file and ran the sample problem. Now, I moving on to Mohr-Coulomb criterion with this [tutorial 2](http://tfel.sourceforge.net/MohrCoulomb.html). There are two code blocks that I saved into MFront files. The first one was in the section "Simplification of the MFront file : use of TFEL/Material/MohrCoulombYieldCriterion.hxx file" and other one is in the section "Simplification of the MFront file : use of the StandardElastoViscoPlasticity brick". Both return this error when compiling:

> In file included from ~/codes/tfel/master/install-661d98cec3cd9c6059b31210be1dbe72e447f11d/include/TFEL/Material/MohrCoulombYieldCriterion.hxx:169,
from …/include/TFEL/Material/MohrCoulomAbboSloan3BehaviourData.hxx:40,
from …/include/TFEL/Material/MohrCoulomAbboSloan3.hxx:33,
from MohrCoulomAbboSloan3-generic.cxx:27:
~/codes/tfel/master/install-661d98cec3cd9c6059b31210be1dbe72e447f11d/include/TFEL/Material/MohrCoulombYieldCriterion.ixx: In lambda function:
~/codes/tfel/master/install-661d98cec3cd9c6059b31210be1dbe72e447f11d/include/TFEL/Material/MohrCoulombYieldCriterion.ixx:110:56: error: ‘sin_angle’ was not declared in this scope
110 | const auto term2 = sign * p.sin_lodeT + isqrt3 * sin_angle * p.cos_lodeT;
| ^~~~~~~~~
compilation terminated due to -Wfatal-errors.
make: *** [Makefile.mfront:36: MohrCoulomAbboSloan3-generic.o] Error 1
callMake: can’t build target ‘all’
libraries building went wrong
Command was: make -C src -f Makefile.mfront all -s

Anyone who had the same problem? Any solution?

Thanks!

EDIT: I stitched the codes in the “Implementation” section of tutorial 2 to create the most explicit Mfront file without the simplifications in the next sections. I compiled it successfully but I have to include “@Parameter pi = 3.14159265359;” and “constexpr auto local_zero_tolerance = real(1e-14);” as they were missing.

**Comment by thelfer** [comment]
Hi @gitable05,

Thanks for you interest to `MFront`.

Would you attach you implementations ?

Best,
Thomas 

BTW, did you used mgis.fenics (<https://thelfer.github.io/mgis/web/mgis_fenics.html>)

**Comment by thelfer** [comment, marked as answer]
@gitable05.

Please find attached three implementations. They all compile with `TFEL-3.4.x` although the third one may fail if `C++-17` support is enabled (See Issue  #80).

For `TFEL-4.x` versions, it seems that PR #5  messed things up and broke the `MohrCoulombStressCriterion.ixx` file. We will treat this in Issue #81. Sorry for the inconvenience.

All in all, those implementations are not robust as the normal may strongly oscillate between two iterations, so it need to be slightly modified using a small unpublished trick. Just send me an email at tfel-contact@cea.fr, and I'll send you a proper implementation next week.

[MohrCoulomb.zip](https://github.com/thelfer/tfel/files/7870623/MohrCoulomb.zip)

**Reply by gitable05** [reply]
@thelfer 

Thank you for the response and the files.

> BTW, did you used mgis.fenics (https://thelfer.github.io/mgis/web/mgis_fenics.html)

I am yet to work on this but I will sure look at mgis.fenics. My plan is to first modify the tutorial with von Mises + MFront to instead have Mohr-Coulomb model. Hopefully, I can get that to work and then base my intended project on that. 

Best.

**Reply by thelfer** [reply]
Again, the files that I sent you are useful for setting up your simulation. In practice, you'll need the modified ones.

**Reply by thelfer** [reply]
@gitable05. For your information, Issue #81 is now closed. Let me know if you managed to use the files that I sent you.

**Reply by gitable05** [reply]
@thelfer Yes, the file from OpenGeoSys works. I am also using mgis.fenics and encountered no issues yet.

Thanks!

**Reply by thelfer** [reply]
@gitable05 Glad to hear that. If possible, do not hesitate to send me a nice picture for the TFEL/MFront twitter account with a little caption and credit.

**Comment by s1966b** [comment]
@thelfer 
Ref: Our email discussion and your kind reply dated 25/04/2023

Dear Thomas,
I apologise for my late response. 

1. As you advised, I tried the command 'tfel-config --includes' (without quotes) which returned the following:
bash: tfel-config: command not found

2. After that I exported the TFEL-HOME variable as advised as follows:
 export TFELHOME=/mfront340
3. Then I retried the same command. This time also it returned the same, i.e.,
 bash: tfel-config: command not found.
For your convenience I am pasting here my original mail content to you for your kind perusal:
=====================================================================================(mail content begins)
a) I have mounted the directory /container/opt/salome_meca/Salome-V2021-s9/prerequisites/Mfront-TFEL340/ in a mount directory "/mfront340" in my singularity shell inside the container.
b) Then I have exported the paths as follows:
===========================================================================
Singularity> export PATH=/mfront340/bin:$PATH
Singularity> export LD_LIBRARY_PATH=/mfront340/lib:$LD_LIBRARY_PATH
Singularity> export LD_INCLUDE_PATH=/mfront340/include/TFEL-3.4.0:$LD_INCLUDE_PATH
===========================================================================
c) I thought that, the include directives written in different files such as "StressCriterion.hxx" etc. are fulfilled by this include path.
d) The directory structure for the mounted dir "/mfront340/" in the container is as follows:
d.1  /mfront340/include/TFEL-3.4.0/MFront/BehaviourBrick/
d.2  /mfront340/include/TFEL-3.4.0/TFEL/Material/ etc.
====================(mail content ends)
I would be grateful if you could further advise me on the matter so that I can compile the Mohr-Coulomb mfront files obtained from you.

Best regards and Thanks for your valuable time.

Sukumar

**Reply by thelfer** [reply]
There are two possibilities for `bash: tfel-config: command not found`:

1. The `PATH` are not set correctly
2. The version number is appended to the executable, so you shall use something like `tfel-config-3.4.0`. In your terminal, just type  `tfel-conf` and use the completion to see if something is available

**Reply by s1966b** [reply]
Dear Thomas
Thanks for your reply. I got the following output
``
Singularity> tfel-c
tfel-check-3.4.0       tfel-config-3.4.0      tfel-config-bin-3.4.0
Singularity> tfel-config-3.4.0 --includes
-I/home/salome/workspace/SALOME-MECA_V2021_PLATFORM/3/salome/prerequisites/install/Mfront-TFEL340/include/TFEL-3.4.0 
Singularity> 
``
Please advise how should I proceed now.

Best regards and thanks.

Sukumar

**Reply by thelfer** [reply]
Does `TFELHOME` affects the output of `tfel-config --includes` ?

**Reply by s1966b** [reply]
Unfortunately not. I got the following, i.e. same output from the 'tfel-config-3.4.0 --includes'.
`Singularity> export TFELHOME=/mfront340
Singularity> tfel-config-3.4.0 --includes
-I/home/salome/workspace/SALOME-MECA_V2021_PLATFORM/3/salome/prerequisites/install/Mfront-TFEL340/include/TFEL-3.4.0`

Hope this helps to find the issue.

Thanks and regards.

Sukumar

**Reply by thelfer** [reply]
That's unexpected.  The code did not change in 3.4.x (compared to 3.4.0) and this works:

```bash
$ tfel-config --includes
-I/home/th202608/codes/tfel/tfel-3.4.x/install-python-3.7/include 
$ export TFELHOME=/tmp
$ tfel-config --includes
-I/tmp/include 
```

**Comment by s1966b** [comment]
Is the singularity container is an issue? 
I had put this question in code-aster forum also, but did not get any advice relevant to the problem.
Should I try the latest container by EDF? 

Regards

**Reply by thelfer** [reply]
I'll give the singularity container a try this evening

**Comment by s1966b** [comment]
Thank you very much Thomas. 

In fact I am using the container with code-aster 15.4 MPI version available at 
"https://github.com/emefff/Code-Aster-MPI-in-Singularity-of-SM2021"
which is based on EDF container of Salome-meca-2021.

Would be waiting to hear from you.
Best regards and Thanks a lot for your time and patience.

Sukumar

**Reply by thelfer** [reply]
Ok. Apparently, the `code_aster` team replaced `tfel-config-3.4.0` by a script that sets the `TFELHOME` variable internally. That's why setting `TFELHOME` has no effect. However, everything seems set correctly by default in the container:

```bash
$ tfel-config-3.4.0 --include-path
/opt/salome_meca/Salome-V2021-s9/prerequisites/Mfront-TFEL340/include/TFEL-3.4.0 
$ ls $(tfel-config-3.4.0 --include-path)
```

**Comment by s1966b** [comment]
Dear Thomas,

Thank you very much for your time and advice.

1. In my unofficial container, the tfel-config --include-path is different and is a non existing directory. I think, that's where the problem lies. 
2. I think, I will need to download the latest container from EDF and will need to compile code-aster MPI myself.  Due to pre-occupations, I will be able to do it possibly in the next month only.
3. I will do the same and try again.
4. Will request your help if needed. 

With warm regards and Best wishes

Sukumar
